### PR TITLE
lua: highlight the active parameter

### PR DESCRIPTION
**Pull request by urandom**
This is for #14217 and #14444

I wanted to somehow highlight the active parameter in the signature help popup. Currently I'm using markdown's `**` symbols to bold the active parameter, though I don't really like this approach. While it seems to work when used with the `ray-x/lsp_signature.nvim` plugin (the markdown is parsed and the active parameter appears in bold and different color), when using the built-in `buf.signature_help` function, the markdown is not processed at all. Is there a way to somehow use a vim-specific escape sequence for a highlight, if such a thing is possible? I'm also not entirely sure whether I'm using the utf16 helper functions correctly.

I also try to support multiple signatures, for languages that support function overloading. I wasn't really able to test this feature, as the only languages I know that support overloading (java, kotlin, python) either don't seem to support the `signatureHelp` request in their language servers or only return the first signature. In theory, this should display all signatures in a single popup, with the active one appearing first. This is what IntelliJ is doing, and I think it's a better approach than requiring user interaction like vscode. In the same light, I also removed the documentation, as that makes the popup rather huge, and I find it distracts from it's task of showing the signature itself (and documentation can be had in the hover action anyway). For this portion, I wonder if it is possible to to know beforehand whether the popup will appear below or above the cursor? Showing the active signature first currently only works if the popup appears below the cursor. However, if the popup appears below, then the active one will be the furthest one from view, which isn't ideal. Or if that's not possible, perhaps there's a way to add a highlight to just the line that contains the active signature?

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -914,23 +936,6 @@ function M.make_floating_popup_options(width, height, opts)
   }
 end
 
-local function _should_add_to_tagstack(new_item)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:917]: Why are these changes present

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -939,33 +944,22 @@ function M.jump_to_location(location)
   -- location may be Location or LocationLink
   local uri = location.uri or location.targetUri
   if uri == nil then return end
+  local bufnr = vim.uri_to_bufnr(uri)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:947]: Again, this seems out of scope for this PR

File: runtime/lua/vim/lsp/util.lua [hunk 3]
```diff
@@ -822,42 +867,19 @@ function M.convert_signature_help_to_markdown_lines(signature_help)
...
-  if signature.parameters and #signature.parameters > 0 then
-    local active_parameter = signature_help.activeParameter or 0
-    -- If the activeParameter is not inside the valid range, then clip it.
-    if active_parameter >= #signature.parameters then
-      active_parameter = 0
-    end
-    local parameter = signature.parameters[active_parameter + 1]
-    if parameter then
-      --[=[
-      --Represents a parameter of a callable-signature. A parameter can
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:838]: Why did you delete the documentation?

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: I must've messed up the rebase, fixed.

File: runtime/lua/vim/lsp/util.lua (hunk 2 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: ditto. fixed

File: runtime/lua/vim/lsp/util.lua (hunk 3 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: i thought it was part of the todo. i brought it back

**Comment by urandom** [comment]
I did some more experiments and found out that if I force the the popup's filetype to be "markdown", then the concealing of certain characters applies, and that hides the `**` characters, leaving only the bold style:

Screenshot from 2021-04-30 17-18-42

Unfortunately, any characters that are part of the markdown, like `_` and `*` will mess up the highlighting, and both of these are rather common in programming languages.

I also experimented with rebuilding the label so that it only includes the parameters:

Screenshot from 2021-04-30 17-16-13

This reduces the verbosity of the popup, since it no longer needs to repeat the function name and return parameters, and is more or less what IntelliJ does. I've set this up to be behind an `parametersOnly` option (like the `border` option that already exists). Finally, I tried using the _backtick_ character to hopefully stop any wild `_` and `*` characters from breaking the highlighting. That somewhat works, but unfortunately, the _backtick_ characters aren't concealed like the `**` ones, and if the _backtick_ -ed also happens to be the active one, thus being surrounded by `**` as well, then any stray `*` also kind of break the syntax highlighting. All in all, not really ideal.

**Comment by urandom** [comment]
Just a thought, but perhaps if we can't easily insert actual highlight groups in a given text (similar to shell color escape sequences), perhaps it would be easier to create a custom filetype that conceals a certain string pair that would be very rare in the wild, while also applying a style to the text between the pair.

**Comment by akinsho** [comment]
@urandom I'm far from an expert but interested in this feature generally, have you considered using `extmarks` that you can then style? see `:h nvim_buf_set_extmark`? I haven't really read through the code properly for what you are doing but seems in theory like a good fit maybe.

**Comment by urandom** [comment]
@akinsho 

I've never used this before, but from what I understand after reading the help text, I would essentially have to produce the simple text for the popup, and a table that holds the lines and cols of each region I want to highlight. Then I would need to modify `util.open_floating_preview` to accept such a table.  After that function has populated the popup with the lines, I would then need to make it call this `nvim_buf_set_extmark` function with a singleton namespace a number of times, passing the coordinates and the hl information to it?

File: runtime/lua/vim/lsp/handlers.lua [hunk 4]
```diff
@@ -333,25 +333,25 @@ M['textDocument/implementation'] = location_handler
...
     print('No signature help available')
     return
   end
-  local syntax = api.nvim_buf_get_option(bufnr, 'syntax')
-  local p_bufnr, _ = util.focusable_preview(method, function()
-    return lines, util.try_trim_markdown_code_blocks(lines), config
+  util.focusable_preview(method, function()
+    return lines, "markdown", config
   end)
-  api.nvim_buf_set_option(p_bufnr, 'syntax', syntax)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/handlers.lua:354]: This is going to remove syntax highlighting, which was a feature request that I closed.

**Comment by mjlbach** [comment]
I don't agree with this approach, I'd recommend using a custom highlight group rather than overriding the syntax rules. We had a lot of people ask for syntax highlighting, which by converting to markdown/inserting ** to trigger the bold conceal rule, you're removing.

File: runtime/lua/vim/lsp/handlers.lua (hunk 4 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua]: @mjlbach  I reverted this change. Will experiment with `nvim_buf_set_extmark`, of perhaps if you can explain what a custom highlight group is and how it can be used

**Comment by urandom** [comment]
Using the extmark api really fixed the issues around highlighting the active parameter. Now, the signature help popup looks like like this with `hi def LspSignatureParamActive gui=bold`:

Screenshot 2021-05-01 at 12 25 15

When just showing params, it looks like:

Screenshot 2021-05-01 at 12 24 35

I also noticed that at least with go, the existing code for setting the syntax didn't actually work, since the original buffer has an empty `syntax` option, so I made it fall back to `filetype` in case `syntax` isn't set.

This change now uses the following highlights: `LspSignatureActive` - in case someone wants to highlight the active signature. There's the corresponding `LspSignatureInactive` for highlighting the rest of the overloaded signatures for a function. And there's `LspSignatureParamActive` for highlighting the active parameter.

**Comment by mjlbach** [comment]
I don't recommend falling back to filetype, as that will result in the language server being attached to the floating window by lspconfig

**Comment by urandom** [comment]
> I don't recommend falling back to filetype, as that will result in the language server being attached to the floating window by lspconfig

I'm falling back to `filetype` for the value to be set by `syntax`, not for setting the actual `filetype`

File: runtime/lua/vim/lsp/handlers.lua [hunk 5]
```diff
@@ -333,25 +335,41 @@ M['textDocument/implementation'] = location_handler
 ---     - border:     (default=nil)
 ---         - Add borders to the floating window
 ---         - See |vim.api.nvim_open_win()|
+---     - parametersOnly: (default=false)
+---         - Display only a list of parameters
+---     - suppressPrint: (default=false)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/handlers.lua:340]: This option seems out of scope of this PR

File: runtime/lua/vim/lsp/util.lua [hunk 6]
```diff
@@ -799,12 +798,116 @@ function M.convert_input_to_markdown_lines(input, contents)
...
+  return label, nil
+end
+
 --- Converts `textDocument/SignatureHelp` response to markdown lines.
 ---
 --@param signature_help Response of `textDocument/SignatureHelp`
 --@returns list of lines of converted markdown.
 --@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_signatureHelp
-function M.convert_signature_help_to_markdown_lines(signature_help)
+function M.get_signature_help_with_highlights(signature_help, parametersOnly)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:910]: We use snakecase everywhere else

File: runtime/lua/vim/lsp/util.lua [hunk 7]
```diff
@@ -822,43 +926,30 @@ function M.convert_signature_help_to_markdown_lines(signature_help)
...
-        --The human-readable doc-comment of this parameter. Will be shown
-        --in the UI but can be omitted.
-        documentation?: string | MarkupContent;
-      }
-      --]=]
-      -- TODO highlight parameter
-      if parameter.documentation then
-        M.convert_input_to_markdown_lines(parameter.documentation, contents)
+
+  for i = 1,#signature_help.signatures,1 do
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:941]: formatting (spaces), also is the last 1 necessary? you're iterating by one anyways

File: runtime/lua/vim/lsp/handlers.lua [hunk 8]
```diff
@@ -7,6 +7,8 @@ local buf = require 'vim.lsp.buf'
 
 local M = {}
 
+local ns = api.nvim_create_namespace("lsp/popup")
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/handlers.lua:10]: I'd recommend making `ns` more descriptive, and `lsp/popup` is ambiguous, maybe `lsp/signature_help`

File: runtime/lua/vim/lsp/handlers.lua [hunk 9]
```diff
@@ -333,25 +335,41 @@ M['textDocument/implementation'] = location_handler
...
   lines = util.trim_empty_lines(lines)
   if vim.tbl_isempty(lines) then
-    print('No signature help available')
+    if not config.suppressPrint then
+      print('No signature help available')
+    end
     return
   end
   local syntax = api.nvim_buf_get_option(bufnr, 'syntax')
+  if syntax == "" then
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/handlers.lua:361]: I'm still not sure why syntax is null for you, it's not for me on go files, and I'd want to understand why before merging this change.

File: runtime/lua/vim/lsp/util.lua [hunk 10]
```diff
@@ -799,12 +798,116 @@ function M.convert_input_to_markdown_lines(input, contents)
   return contents
 end
 
+local function parameter_range_in_signature(signature, active_parameter)
+  local from
+  local to = 0
+  for i = 0,#signature.parameters-1,1 do
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/util.lua:804]: Same comment, spaces and the iterate interval is unnecessary

File: runtime/lua/vim/lsp/util.lua (hunk 7 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua:941]: I've no idea. Lua isn't a language I'm skilled at. Can I omit it?

File: runtime/lua/vim/lsp/handlers.lua (hunk 8 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua:10]: I don't really know what that namespace is used for. I'll change names to something more descriptive.

File: runtime/lua/vim/lsp/handlers.lua (hunk 9 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua:361]: No idea. I just know that it is, and I've never actually seen highlighted signature popups

File: runtime/lua/vim/lsp/handlers.lua (hunk 8 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp/handlers.lua:10]: lsp_highlight_ns for example

File: runtime/lua/vim/lsp/handlers.lua (hunk 5 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua:340]: It seemed like a low hanging fruit, and I saw that the lsp_signature plugin is redefining the same handler just to avoid the prints. Should I revert that particular change?

File: runtime/lua/vim/lsp/handlers.lua (hunk 9 above)
Comment (mjlbach) [reply runtime/lua/vim/lsp/handlers.lua:361]: Something may be wrong with your config then: 

image

**Comment by mjlbach** [comment]
A couple comments:

1. I think this is great, and I'm excited to see this feature implemented, as it's definitely something that gets brought up a lot, so thank you for working on this!

2. This doesn't fully close #14217, as we'd need to provide an interface for navigating between signature items. That doesn't have to be done in this PR, but you should remove the 'closes #14217' tag if you don't intend to do it in this PR

3. The extended marks approach is the way to go, thank you for making that work.

4. I need to find some time to review the PR in more technical detail, I'm not sure if there's a simpler way to match the active signature.

File: runtime/lua/vim/lsp/handlers.lua (hunk 8 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua]: fixed

File: runtime/lua/vim/lsp/util.lua (hunk 7 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: fixed

File: runtime/lua/vim/lsp/util.lua (hunk 10 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: fixed

File: runtime/lua/vim/lsp/util.lua (hunk 6 above)
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua]: fixed

**Comment by urandom** [comment]
> * 
> 
> * This doesn't fully close #14217, as we'd need to provide an interface for navigating between signature items. That doesn't have to be done in this PR, but you should remove the 'closes #14217' tag if you don't intend to do it in this PR

Perhaps I'm misunderstanding #14217, but I thought this would fix it by virtue of displaying all signatures (with the active one being on top). This approach does not require any user interaction, so no need to be able to navigate between them. Would that not be the case?

**Comment by mjlbach** [comment]
Here is the motivating reddit comment by that issue's OP https://old.reddit.com/r/neovim/comments/mdj3ct/using_nvim_integrated_lsp_i_cant_get_signature/gs9ophu/

I haven't tested yet, but functions with many args may get messy if it's in the same window

File: runtime/lua/vim/lsp/handlers.lua (hunk 9 above)
Comment (urandom) [reply runtime/lua/vim/lsp/handlers.lua:361]: That's the most likely explanation, or some plugin. `verbose set syntax?` just returns `Last set from Lua`. Though this is the first time I've had a problem with the lack of a set `syntax` option.

Comment (zacmar) [reply runtime/lua/vim/lsp/handlers.lua:361]: I have also been experiencing something similar (see also [https://github.com/ray-x/lsp_signature.nvim/issues/13](https://github.com/ray-x/lsp_signature.nvim/issues/13)). Turns out, if I enable treesitter highlighting, I lose the markdown highlighting in the signature help.
`init.lua`
```lua
vim.cmd('packadd! nvim-lsp')
vim.cmd('packadd! nvim-treesitter')
vim.api.nvim_set_keymap("n", "gk", "<cmd>lua vim.lsp.buf.signature_help()<CR>", {})
require'lspconfig'.gopls.setup{}
require'nvim-treesitter.configs'.setup {
    highlight = {
        enable = true,
    },
}
```
`nvim -u init.lua test.go`
2021-05-02-122439_859x195_scrot
`init_.lua`
```lua
vim.cmd('packadd! nvim-lsp')
vim.api.nvim_set_keymap("n", "gk", "<cmd>lua vim.lsp.buf.signature_help()<CR>", {})
require'lspconfig'.gopls.setup{}
```
`nvim -u init_.lua test.go`
2021-05-02-122722_837x165_scrot
Not sure what the problem is here, of if this is intendet behavior. But I would be interested in how this is resolved.

Comment (ray-x) [reply runtime/lua/vim/lsp/handlers.lua:361]: I turn it off syntax in my init.lua because I am using treesitter.
That can save 15ms~200ms when loading.   As syntax highlighting can be slow and if you need treesitter to render it twice, makes loading slower.
I can use treesitter to highlight the signature help. The only problem is it is missing `func` before Println and treesitter will complaining about it.

File: runtime/lua/vim/lsp/util.lua [hunk 11]
```diff
@@ -822,43 +926,30 @@ function M.convert_signature_help_to_markdown_lines(signature_help)
   if not signature then
     return
   end
-  vim.list_extend(contents, vim.split(signature.label, '\n', true))
-  if signature.documentation then
-    M.convert_input_to_markdown_lines(signature.documentation, contents)
+
+  local active_parameter = signature.activeParameter or signature_help.activeParameter or 0
```
[the remark this answers is not in the document]
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua:930]: do you know what -1 would mean in that context? No active signature?

[the remark this answers is not in the document]
Comment (ray-x) [reply runtime/lua/vim/lsp/util.lua:930]: I forget why it return -1 when everyone else return 0...

[the remark this answers is not in the document]
Comment (urandom) [reply runtime/lua/vim/lsp/util.lua:930]: If -1 is meant as a value that indicates no active parameter, then the code will not fail, but instead not highlight any parameter, which would be the intended behaviour.

**Comment by mjlbach** [comment]
Nope, this is just a medium-ish PR coming very soon before 0.5, and I've been busy with real stuff. It's easier for me to review/approve bug fixes/add completely new behavior than to review/merge PRs that change functionality/default behavior.

**Comment by mjlbach** [comment]
The current implementation may need some modifications to implement more of the functionality implemented PR, but this PR as it stands can be closed.

**Comment by urandom** [comment]
@mjlbach 
Should I do anything to bring any of the additional features to the current implementation, or should I scrap the whole thing?

**Comment by mjlbach** [comment]
Sure! You can coordinate with @folke who can also help review

**Comment by folke** [comment]
We had some discussion on whether we should show all signatures, versus only the active one and the consensus was that it makes most sense to just show the active signature.

Most of your points in the description about rendering should be taken into account, since we're using the stylized markdown to render the signature, docs and possible docs for the active parameter.

Is there anything else in this PR that you think might be beneficial to further improve signature help?

**Comment by urandom** [comment]
@folke, @mjlbach 

I guess there are four possible areas I am personally interested in improving. Three are very low hanging:
1: remove or have an option to not print out a message when no signature is available (for when the signature is used with an autocommand)
2: "Align" the popup with the active parameter (this one is rather easy, since all there needs to be done is pass the active parameter offset as an offset to the popup). Not as good as keeping the popup visible at all times while writing the input parameters, but it definitely improves things.
3: Have an option to not show the signature help. For me this would be quite useful, as it would remove the clutter from the popup, since I'm only interested in the signature (and the help is available through hover anyway)

The fourth one is a bit more involved:
4: Have an option to only display the parameters, as opposed to the full signature. The full signature seems to consist of the function name and return parameters, but at least for me, I only care about the input parameters at this point, since that is what I am writing. This would further reduce clutter for me, leaving only the important bits (the input parameters) in the popup.
